rpcs3/Emu/RSX/Overlays/overlay_message.cpp

```diff
@@ -6,88 +6,178 @@ namespace rsx
+			m_fade_in_animation.active = true;
+
+			m_fade_out_animation.current = color4f(1.f);
+			m_fade_out_animation.end = color4f(0.f);
+			m_fade_out_animation.duration = 2.f;
+			m_fade_out_animation.active = true;
+
+			back_color = color4f(0.25f, 0.25f, 0.25f, 0.85f);
+
+			if (icon)
+			{
+				m_icon = std::move(icon);
+				m_icon->set_pos(m_text.x + m_text.w + 8, m_text.y);
+
+				set_size(m_margin + m_text.w + m_icon->w + m_margin, m_margin + std::max(m_text.h, m_icon->h) + m_margin);
+			}
+			else
+			{
+				set_size(m_text.w + m_margin + m_margin, m_text.h + m_margin + m_margin);
+			}
 		}
-		template message_item::message_item(std::string msg_id, u64, std::shared_ptr<atomic_t<u32>>);
-		template message_item::message_item(localized_string_id msg_id, u64, std::shared_ptr<atomic_t<u32>>);
+		template message_item::message_item(std::string msg_id, u64, std::shared_ptr<atomic_t<u32>>, std::unique_ptr<overlay_element>);
+		template message_item::message_item(localized_string_id msg_id, u64, std::shared_ptr<atomic_t<u32>>, std::unique_ptr<overlay_element>);
 
 		u64 message_item::get_expiration() const
 		{
 			// If reference counting is enabled and reached 0 consider it expired 
 			return m_refs && *m_refs == 0 ? 0 : m_expiration_time;
 		}
 
-		compiled_resource message_item::get_compiled()
+		bool message_item::text_matches(const std::u32string& text) const
 		{
-			if (!m_processed)
+			return m_text.text == text;
+		}
+
+		void message_item::set_pos(u16 _x, u16 _y)
+		{
+			rounded_rect::set_pos(_x, _y);
+			m_text.set_pos(_x + m_margin, y + m_margin);
+
+			if (m_icon)
 			{
-				return {};
+				m_icon->set_pos(m_icon->x, m_text.y);
 			}
+		}
 
-			auto cr = m_text.get_compiled();
-			m_fade_animation.apply(cr);
+		compiled_resource& message_item::get_compiled()
+		{
+			auto& current_animation = m_fade_in_animation.active
+				? m_fade_in_animation
+				: m_fade_out_animation;
 
-			return cr;
+			if (!m_processed || !current_animation.active)
+			{
+				compiled_resources = {};
+				return compiled_resources;
+			}
+
+			// Disable caching
+			is_compiled = false;
+
+			compiled_resources = rounded_rect::get_compiled();
+			compiled_resources.add(m_text.get_compiled());
+			if (m_icon)
+			{
+				compiled_resources.add(m_icon->get_compiled());
+			}
+
+			current_animation.apply(compiled_resources);
+			return compiled_resources;
 		}
 
-		void message_item::update(usz index, u64 time)
+		void message_item::update(usz index, u64 time, u16 y_offset)
 		{
 			if (m_cur_pos != index)
 			{
 				m_cur_pos = index;
-				m_text.set_pos(10, static_cast<u16>(index * 18));
+				set_pos(10, y_offset);
+			}
+
+			if (!m_processed)
+			{
+				m_expiration_time = get_expiration_time(m_visible_duration);
 			}
 
-			if ((m_expiration_time - time) < 2'000'000)
+			if (m_fade_in_animation.active)
+			{
+				m_fade_in_animation.update(rsx::get_current_renderer()->vblank_count);
+			}
+			else if (time + 2'000'000 > m_expiration_time)
```

Comment:
you can probably use the m_fade_out_animation.duration to calculate this

Reply:
took the opportunity to decrease fadeout to 1s as well